src/PrMerger.js

```diff
@@ -43,12 +44,31 @@ class PrMerger {
     // there is a PR still-in-merge.
     async runStep() {
         Logger.info("runStep running");
-        const currentContext = await this._current();
-        if (currentContext && !(await this._finishContext(currentContext)))
-            return true; // still in-process
 
-        const prList = await this._getPRList();
-        this._logPRList(prList);
+        let prList = await this._getPRList();
+
+        let currentContext = await this._current();
+
+        if (currentContext && Config.guardedRun()) {
+            const clearedForMerge = await this._clearedForMerge(currentContext.number());
+            if (!clearedForMerge && prList.length && prList[0].clearedForMerge) {
+                Logger.info("will try to give up the not-cleared-for-merge PR" + currentContext.number() +
+                        " in favor of a cleared-for-merge PR" + prList[0].number);
+                // Give only other PRs a chance to start.
+                // Without removing current PR from the list, we can get into
+                // a loop of useless restarting it from scratch.
+                prList = prList.filter(pr => pr.number !== currentContext.number());
+                currentContext = null;
+            }
+        }
+
+        if (currentContext) {
+            if (!(await this._finishContext(currentContext)))
+                return true; // still in-process
+            // refresh the list
+            prList = await this._getPRList();
```

Comment:
I do not like the complexity this "cache update" introduces. After we finished merged something, can we just return and start from scratch instead?

Reply:
We could "start from scratch" but this would make another complexity: how the upper-level code (RepoMerger::run()) would know that we want to rerun? We would have to manipulate with RepoMerger::rerun flag in low-level code, which is not good (IIRC we rejected such manipulation some time ago).